solr/jetty/src/java/org/apache/solr/jetty/ShrinkManager.java

```diff
@@ -0,0 +1,145 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.apache.solr.jetty;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.eclipse.jetty.util.NanoTime;
+
+/**
+ * <p>Encapsulates logic for shrinking {@link QueuedThreadPool}. Tracks pool-level idle baseline
+ * timestamps as a deque. Upon idle, threads push to the head of the deque by calling {@link #onIdle()},
+ * and when an idle thread becomes non-idle (i.e., runs a job), it pops from the head of the deque
+ * by calling {@link #onBusy()}.
+ *
+ * <p>Shrinking proceeds by polling the <i>tail</i> of the deque via {@link #pollIdleShrink(long, long)};
+ * the pool thus determines eligibility to shrink based on the longest-standing idle capacity at the pool
+ * level.
+ *
+ * <p>The repeated lifecycle methods {@link #onIdle()} and {@link #onBusy()} return <code>true</code>
+ * and <code>false</code> (respectively) as a convenience, indicating whether the calling thread is
+ * registered with this {@link ShrinkManager} following the associated method call. It is the
+ * responsibility of the calling thread to ensure (e.g., via conditional call to {@link #prune()}
+ * in a <code>finally</code> block) that any outstanding entries in this {@link ShrinkManager} are
+ * removed if the calling thread exits unexpectedly (i.e., throws an exception).
+ */
+public class ShrinkManager
```

Comment:
QQ. In every idleTimeoutDecay interval we shrink one thread now? earlier it used to happen for every idleTimeout interval?

Reply:
Sure. let me take a example for it. 
1. minthreads =1
2. idletimeout= 60 seconds
3. idleTimeoutDecay (i want 5 seconds after quick burst of requests - not clear how I will configure that)

===
let's say 10 threads created. Now how threads will shrink by time
65,70,75,80,85,90,95,100, 105 - (1 thread will remain)
is my understanding correct?

Reply:
I think the confusion stems from the fact that `idleTimeoutDecay` is ultimately a _number of threads_, not an amount of time. So If you want _all threads_ to shrink immediately after `idleTimeout` has passed, you'd set `idleTimeoutDecay` equal to `maxThreads`. Could set it higher, but it shouldn't make a difference practically speaking. So `idleTimeoutDecay=1` would be equivalent to current behavior, where 1 thread is allowed to expire per `idleTimeout` interval.

Reply:
Yes, first end user doesn't expect behavior of current implementation of `idletimeout`. Now, it feels to me this new parameter will complicate the things more. 

I think it would be simple if we can consider new parameter as timeout decay
1. default timeout decay will be `idletimeout`
2. Easier to understand
3. More importantly, there is `minthreads` parameter to keep minimum threads alive.
4. Also, it may simplify the implementation.

Reply:
I'm not sure what the desired behavior you have in mind is?

Reply:
I'm thinking of following two things
1. keep current behavior as it is
2. if thread is idle for `idletimeout` then shrink it (decaytimeout =0). That's what most users(all) expects
3. In-between, decay every 5 seconds as described in above example.

Reply:
I was looking [this](https://github.com/eclipse/jetty.project/commit/17391afbc75a57bb606660e3eef9de1979d8c3f8) commit. That seems bit simple implementation